clients/src/main/java/org/apache/kafka/clients/consumer/StickyAssignor.java

```diff
@@ -374,19 +376,37 @@ public Subscription subscription(Set<String> topics) {
         if (memberAssignment == null)
             return new Subscription(new ArrayList<>(topics));
 
-        return new Subscription(new ArrayList<>(topics),
-                serializeTopicPartitionAssignment(new ConsumerUserData(memberAssignment, Optional.of(generation))));
+        return buildSubscription(topics);
     }
 
     @Override
     public String name() {
         return "sticky";
     }
 
+    List<TopicPartition> memberAssignment() {
+        return memberAssignment;
+    }
+
     int generation() {
         return generation;
     }
 
+    // This method may be overriden by a subclass that follows sticky assignment but encodes Subscription/UserData differently
+    StickyUserData subscriptionToStickyUserData(Subscription subscription) {
+        ByteBuffer userData = subscription.userData();
+        if (userData == null || !userData.hasRemaining())
+            return  null;
+
+        return deserializeTopicPartitionAssignment(userData);
+    }
+
+    // This method may be overriden by a subclass that encodes Subscription differently
+    Subscription buildSubscription(Set<String> topics) {
```

Comment:
Should we define abstract class called `AbstractStickyAssignor` and make `buildSubscription` and `subscriptionToStickyUserData` abstract functions?

Reply:
See above 😄